From 611e34b1c6a8a18c51e8c4477d41fd97d5329fcb Mon Sep 17 00:00:00 2001
From: Charles Keepax <ckeepax@opensource.wolfsonmicro.com>
Date: Fri, 30 Aug 2013 16:15:54 +0100
Subject: [PATCH 033/351] ASoC: wm_adsp: Allow DSP time to set host_buf_ptr

It seems host_buf_ptr is set dynamically on the DSP and if we trigger
the compressed stream too soon, we might read zero. Poll this value a
few times to give the DSP time to set it up.

Change-Id: Ia36abaf4ed001c43818391f6361b963b1b454b44
Signed-off-by: Charles Keepax <ckeepax@opensource.wolfsonmicro.com>
Signed-off-by: Nikesh Oswal <Nikesh.Oswal@wolfsonmicro.com>
---
 sound/soc/codecs/wm_adsp.c |   24 +++++++++++++++++-------
 1 file changed, 17 insertions(+), 7 deletions(-)

diff --git a/sound/soc/codecs/wm_adsp.c b/sound/soc/codecs/wm_adsp.c
index 1df064f..8f5c0d9 100644
--- a/sound/soc/codecs/wm_adsp.c
+++ b/sound/soc/codecs/wm_adsp.c
@@ -2325,7 +2325,7 @@ EXPORT_SYMBOL_GPL(wm_adsp_stream_free);
 int wm_adsp_stream_start(struct wm_adsp *adsp)
 {
 	u32 xm_base, magic;
-	int ret;
+	int i, ret;
 
 	ret = wm_adsp_read_data_word(adsp, WMFW_ADSP2_XM,
 				     ADSP2_SYSTEM_CONFIG_XM_PTR, &xm_base);
@@ -2342,12 +2342,22 @@ int wm_adsp_stream_start(struct wm_adsp *adsp)
 	if (magic != WM_ADSP_ALG_XM_STRUCT_MAGIC)
 		return -EINVAL;
 
-	ret = wm_adsp_read_data_word(adsp, WMFW_ADSP2_XM,
-				     xm_base + WM_ADSP_ALG_XM_PTR +
-				     ALG_XM_FIELD(host_buf_ptr),
-				     &adsp->host_buf_ptr);
-	if (ret < 0)
-		return ret;
+	for (i = 0; i < 5; ++i) {
+		ret = wm_adsp_read_data_word(adsp, WMFW_ADSP2_XM,
+					     xm_base + WM_ADSP_ALG_XM_PTR +
+					     ALG_XM_FIELD(host_buf_ptr),
+					     &adsp->host_buf_ptr);
+		if (ret < 0)
+			return ret;
+
+		if (adsp->host_buf_ptr)
+			break;
+
+		msleep(1);
+	}
+
+	if (!adsp->host_buf_ptr)
+		return -EIO;
 
 	ret = wm_adsp_host_buffer_read(adsp,
 				       HOST_BUFFER_FIELD(low_water_mark),
-- 
1.7.9.5

